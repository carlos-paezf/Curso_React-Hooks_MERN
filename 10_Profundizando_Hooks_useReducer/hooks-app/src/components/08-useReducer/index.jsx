export { TodoApp } from "./TodoApp";
export { TodoList } from "./TodoList";
export { TodoListItem } from "./TodoListItem";
export { TodoAdd } from "./TodoAdd";