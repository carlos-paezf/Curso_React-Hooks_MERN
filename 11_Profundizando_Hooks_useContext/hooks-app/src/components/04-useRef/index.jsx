export { FocusScreen } from './FocusScreen'
export { RealExampleRef } from './RealExampleRef'