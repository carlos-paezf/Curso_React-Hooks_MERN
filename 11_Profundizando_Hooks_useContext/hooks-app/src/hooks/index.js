/* The useFetch hook is a function that takes in a URL and returns a response. 
The useForm hook is a function that takes in a form and returns a form. 
The useCounter hook is a function that takes in a number and returns a number. */
export { useFetch } from "./useFetch";
export { useForm } from "./useForm";
export { useCounter } from "./useCounter";