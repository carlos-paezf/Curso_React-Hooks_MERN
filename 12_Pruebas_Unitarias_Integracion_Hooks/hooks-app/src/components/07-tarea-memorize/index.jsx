export { Hijo } from "./Hijo";
export { Padre } from "./Padre";