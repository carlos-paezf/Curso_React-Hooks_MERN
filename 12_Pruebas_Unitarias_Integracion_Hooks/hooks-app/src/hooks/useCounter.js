import { useState } from "react";

/**
 * @param [initialState=10] - The initial state of the counter.
 * @returns The state, increment, decrement, and reset functions.
 */
export const useCounter = (initialState = 10) => {
    const [state, setState] = useState(initialState);

    const increment = (factor = 1) => setState(state + factor)

    const decrement = (factor = 1) => setState(state - factor)

    const reset = () => setState(initialState)

    return { state, increment, decrement, reset };
};
