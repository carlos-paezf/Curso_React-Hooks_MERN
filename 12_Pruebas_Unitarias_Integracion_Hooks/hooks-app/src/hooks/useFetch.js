import { useEffect, useRef, useState } from "react";

/**
 * @param url - the URL to fetch
 * @returns The state object.
 */
export const useFetch = (url) => {

    const isMounted = useRef(true);

    const [state, setState] = useState({
        data: null,
        loading: true,
        error: null
    });

    useEffect(() => {
        return () => {
            isMounted.current = false
        }
    }, []);


    useEffect(() => {
        setState({ data: null, loading: true, error: null })
        fetch(url)
            .then(res => res.json())
            .then(data => {
                if (isMounted.current) {
                    setState({
                        data,
                        loading: false,
                        error: null
                    })
                } else {
                    console.log('setState no se llamo')
                }
            })
            .catch(error => {
                setState({
                    data: null,
                    loading: false,
                    error: 'No se pudo cargar la data'
                })
            })
    }, [url]);

    return state
}
