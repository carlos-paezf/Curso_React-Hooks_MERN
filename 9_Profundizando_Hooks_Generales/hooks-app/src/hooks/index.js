export { useFetch } from "./useFetch";
export { useForm } from "./useForm";
export { useCounter } from "./useCounter";