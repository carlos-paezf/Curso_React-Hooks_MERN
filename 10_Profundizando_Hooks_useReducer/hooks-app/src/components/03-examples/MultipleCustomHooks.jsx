import React from 'react';
import { useCounter, useFetch } from '../../hooks';


export const MultipleCustomHooks = () => {

    const { state, increment } = useCounter(1)

    const { data, loading } = useFetch(`https://www.breakingbadapi.com/api/quotes/${state}`)


    const { author, quote } = !!data && data[0]

    return (
        <>
            <h2>Multiple CustomHooks - Breaking Bad Quotes</h2>
            <hr />
            {
                loading
                    ? <div className="alert alert-info text-center">...Loading...</div>
                    : <blockquote className="blockquote">
                        <p className="mb-2">{quote}</p>
                        <footer className="blockquote-footer">{author}</footer>
                    </blockquote>
            }
            <button className="btn btn-dark" onClick={() => increment(1)}>Siguiente Cita</button>
        </>
    );
};
