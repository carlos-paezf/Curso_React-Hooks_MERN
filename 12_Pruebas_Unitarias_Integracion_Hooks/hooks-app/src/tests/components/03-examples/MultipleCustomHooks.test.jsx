import React from 'react';
import { shallow } from 'enzyme'
import { MultipleCustomHooks } from "../../../components/03-examples/"
import { useFetch, useCounter } from '../../../hooks';

jest.mock('../../../hooks/useFetch')
jest.mock('../../../hooks/useCounter')


describe('Pruebas al componente <Multiple Custom Hook />', () => {

    beforeEach(() => {
        useCounter.mockReturnValue({
            state: 10,
            increment: jest.fn(),
            decrement: jest.fn(),
            reset: jest.fn()
        })
    })
    
    /* The useFetch hook is used to fetch data from an API. The useFetch hook returns an object with
    the following properties:
    
    data: The data returned by the API.
    loading: A boolean indicating whether the data is being fetched.
    error: The error returned by the API. */
    test('Debe de mostrarse correctamente', () => {
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        })

        const wrapper = shallow(<MultipleCustomHooks />)

        expect(wrapper).toMatchSnapshot()
    })


    test('Debe de mostrar la informaciÃ³n', () => {
        useFetch.mockReturnValue({
            data: [{
                author: 'Ferrer',
                quote: 'Hola mundo'
            }],
            loading: false,
            error: null
        })

        const wrapper = shallow(<MultipleCustomHooks />)

        expect(wrapper.find('.alert').exists()).toBe(false)
        expect(wrapper.find('.mb-2').text().trim()).toBe('Hola mundo')
        expect(wrapper.find('footer').text().trim()).toBe('Ferrer')
    })
})
