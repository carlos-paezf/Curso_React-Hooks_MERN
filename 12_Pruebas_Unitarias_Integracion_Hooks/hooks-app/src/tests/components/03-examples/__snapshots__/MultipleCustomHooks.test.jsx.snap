// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas al componente <Multiple Custom Hook /> Debe de mostrarse correctamente 1`] = `
<Fragment>
  <h2>
    Multiple CustomHooks - Breaking Bad Quotes
  </h2>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    ...Loading...
  </div>
  <button
    className="btn btn-dark"
    onClick={[Function]}
  >
    Siguiente Cita
  </button>
</Fragment>
`;
